x/stakeibc/keeper/callbacks.go

```diff
@@ -194,69 +194,89 @@ func WithdrawalBalanceCallback(k Keeper, ctx sdk.Context, args []byte, query icq
 	return nil
 }
 
-// get a validator and its index from a list of validators, by address
-func getValidator(validators []*types.Validator, address string) (types.Validator, int64, bool) {
-	for i, v := range validators {
-		if v.Address == address {
-			return *v, int64(i), true
-		}
-	}
-	return types.Validator{}, 0, false
-}
-
 // ValidatorCallback is a callback handler for validator queries.
+//
+// In an attempt to get the ICA's delegation amount on a given validator, we have to query:
+//     1) the validator's internal exchange rate
+//     2) the Delegation ICA's delegated shares
+//  And apply the following equation:
+//     num_tokens = exchange_rate * num_shares
+//
+// This callback from query #1
 func ValidatorExchangeRateCallback(k Keeper, ctx sdk.Context, args []byte, query icqtypes.Query) error {
-	zone, found := k.GetHostZone(ctx, query.GetChainId())
+	hostZone, found := k.GetHostZone(ctx, query.GetChainId())
 	if !found {
-		return fmt.Errorf("no registered zone for chain id: %s", query.GetChainId())
+		errMsg := fmt.Sprintf("no registered zone for queried chain ID (%s)", query.GetChainId())
+		k.Logger(ctx).Error(errMsg)
+		return sdkerrors.Wrapf(types.ErrHostZoneNotFound, errMsg)
 	}
 	queriedValidator := stakingtypes.Validator{}
 	err := k.cdc.Unmarshal(args, &queriedValidator)
 	if err != nil {
-		k.Logger(ctx).Error(fmt.Sprintf("unable to unmarshal queriedValidator info for zone %s, err: %s", zone.ChainId, err.Error()))
-		return err
+		errMsg := fmt.Sprintf("unable to unmarshal queriedValidator info for zone %s, err: %s", hostZone.ChainId, err.Error())
+		k.Logger(ctx).Error(errMsg)
+		return sdkerrors.Wrapf(types.ErrMarshalFailure, errMsg)
 	}
-	k.Logger(ctx).Info(fmt.Sprintf("ValidatorCallback: zone %v queriedValidator %v", zone.ChainId, queriedValidator))
+	k.Logger(ctx).Info(fmt.Sprintf("ValidatorCallback: zone %v queriedValidator %v", hostZone.ChainId, queriedValidator))
 
 	// ensure ICQ can be issued now! else fail the callback
-	valid, err := k.IsWithinBufferWindow(ctx)
+	withinBufferWindow, err := k.IsWithinBufferWindow(ctx)
 	if err != nil {
+		// QUESTION: this is the case where the user submitted the query within the buffer window,
```

Comment:
It looks like the errors that could surface here are (1) int casts gone wrong (2) epoch trackers not being found

Reply:
ah good eye! You're right I read it wrong